libc/stdio_printf.c

```diff
@@ -0,0 +1,3322 @@
+	count = 0;
+
+	vftest(expect, format, &count);
+	TEST_ASSERT_EQUAL_INT(strlen(expect), count);
+	count = 0;
+
+	vtest(expect, format, &count);
+	TEST_ASSERT_EQUAL_INT(strlen(expect), count);
+	count = 0;
+
+	vsntest(expect, format, &count);
+	TEST_ASSERT_EQUAL_INT(strlen(expect), count);
+	count = 0;
+
+	vstest(expect, format, &count);
+	TEST_ASSERT_EQUAL_INT(strlen(expect), count);
+	count = 0;
+}
+
+
+TEST(stdio_printf_cspn, percent)
+{
+	const char *format = "%% %yes % --- % yes %%Lorem%%Ipsum %% Hello";
+	char *expect = "% %yes % yes %Lorem%Ipsum % Hello";
+
+	dtest(format, 0, expect);
+	ftest(format, 0, expect);
+	stdtest(format, 0, expect);
+	sntest(format, 0, expect);
+	stest(format, 0, expect);
+
+	vdtest(expect, format, 0);
+	vftest(expect, format, 0);
+	vtest(expect, format, 0);
+	vsntest(expect, format, 0);
+	vstest(expect, format, 0);
+}
+
+///////////////////////////////////////////////////////////////////////////
+
+TEST_SETUP(stdio_printf_rest)
+{
+	output_file = fopen(PATH, "w+");
+}
+
+
+TEST_TEAR_DOWN(stdio_printf_rest)
+{
+	fclose(output_file);
+	remove(PATH);
+}
+
+
+TEST(stdio_printf_rest, mods_int)
+{
+	int i;
+	const char *format = "%2d %5d %05d %+d %.5d %-10d";
+	const int values[] = { INT_MAX, 0, INT_MIN, -123, 123, 123456789 };
+	char *expect[] = { "2147483647 2147483647 2147483647 +2147483647 2147483647 2147483647", " 0     0 00000 +0 00000 0         ", "-2147483648 -2147483648 -2147483648 -2147483648 -2147483648 -2147483648", "-123  -123 -0123 -123 -00123 -123      ", "123   123 00123 +123 00123 123       ", "123456789 123456789 123456789 +123456789 123456789 123456789 " };
+
+	for (i = 0; i < (sizeof(values) / sizeof(values[0])); i++) {
+
+		dtest(format, values[i] COMMA values[i] COMMA values[i] COMMA values[i] COMMA values[i] COMMA values[i], expect[i]);
+		ftest(format, values[i] COMMA values[i] COMMA values[i] COMMA values[i] COMMA values[i] COMMA values[i], expect[i]);
+		stdtest(format, values[i] COMMA values[i] COMMA values[i] COMMA values[i] COMMA values[i] COMMA values[i], expect[i]);
+		sntest(format, values[i] COMMA values[i] COMMA values[i] COMMA values[i] COMMA values[i] COMMA values[i], expect[i]);
+		stest(format, values[i] COMMA values[i] COMMA values[i] COMMA values[i] COMMA values[i] COMMA values[i], expect[i]);
+
+		vdtest(expect[i], format, values[i], values[i], values[i], values[i], values[i], values[i]);
+		vftest(expect[i], format, values[i], values[i], values[i], values[i], values[i], values[i]);
+		vtest(expect[i], format, values[i], values[i], values[i], values[i], values[i], values[i]);
+		vsntest(expect[i], format, values[i], values[i], values[i], values[i], values[i], values[i]);
+		vstest(expect[i], format, values[i], values[i], values[i], values[i], values[i], values[i]);
+	}
+}
+
+
+TEST(stdio_printf_rest, mods_float)
+{
+	int i;
+	const char *format = "%2f %10f %010f %+f %.5f %-10f";
+	const float values[] = { FLT_MAX, 0, FLT_MIN, -1.23, 1.23, 0.23456789 };
```

Comment:
I think adding there DBL_MAX and DBL_MIN values could be valuable and certainly asserting the significant figures in edge min values.